themes/default/content/blog/url-shortener-cloudflare-worker/index.md

```diff
@@ -0,0 +1,276 @@
+
+```javascript
+const DEFAULT_URL = "https://youtube.com/pulumitv"
+
+const redirects = {
+  "pulumi.tv": {
+    "modern-infrastructure": {
+        to: "https://www.youtube.com/playlist?list=PLyy8Vx2ZoWloyj3V5gXzPraiKStO2GGZw"
+    }
+  },
+};
+
+addEventListener("fetch", async (event) => {
+  event.respondWith(handleRequest(event.request));
+});
+
+async function handleRequest(request) {
+  const requestUrl = new URL(request.url);
+  const domain = requestUrl.host;
+  const path = requestUrl.pathname.substring(1);
+
+  console.log(`Handling request on domain ${domain} for ${path}`);
+
+  if (!(domain in redirects)) {
+    console.log(`Domain '${domain}' not found`);
+    return Response.redirect(DEFAULT_URL);
+  }
+
+  if (path === "") {
+    return Response.redirect(DEFAULT_URL);
+  }
+
+  const paths = redirects[domain];
+
+  if (path in paths) {
+    console.log(`Redirecting too ${paths[path].to}`);
+    return Response.redirect(paths[path].to);
+  }
+
+  console.log(`Path '${path}' not found`);
+  return Response.redirect(DEFAULT_URL);
+}
+```
+
+## Deploying Our Worker
+
+```typescript
+const worker = new cloudflare.WorkerScript(
+  "url-shortener",
+  {
+    name: "url-shortener",
+    content: fs.readFileSync("worker.js").toString(),
+  },
+);
+```
+
+## Where's the Types?!
+
+If you wanted to use TypeScript for some guarantees around the short/long URL objects, you could store them as such:
+
+```typescript
+interface Redirect {
+  to: string;
+}
+
+interface Redirects {
+  [domain: string]: {
+    [path: string]: Redirect;
+  };
+}
+
+const redirects: Redirects = {
+  "pulumi.tv": {
+    "/": {
+      to: "https://youtube.com/pulumitv",
+    },
+    "modern-infrastructure": {
+        to: "https://www.youtube.com/playlist?list=PLyy8Vx2ZoWloyj3V5gXzPraiKStO2GGZw"
+    }
+  },
+};
+```
+
+Then, using local commands, transpile them into CommonJS:
+
+```typescript
+const compileWorker = new command.local.Command("compile-worker", {
+  create:
+    "yarn run tsc >/dev/null && cat ./dist/index.js",
+  dir: "../worker",
+});
+```
+
+You could then use `compileWorker.stdout` as instead of `fs.readFileSync` within the worker script.
+
+## Hooking Up a Domain
+
+Cloudflare Workers allow us to connect a domain name, instead of using the generated `workers.dev` subdomain. To do so with Pulumi, we need to [create]({{< relref "/registry/packages/cloudflare/api-docs/zone/" >}}) or [fetch]({{< relref "registry/packages/cloudflare/api-docs/getzone/" >}}) a Cloudflare Zone.
+
+### Create a Zone(s)
+
+Once we create the `cloudflare.Zone` resource, we also need to create a `cloudflare.ZoneSettingsOverride` resource to ensure that we enable:
+
+- `alwaysUseHttps: "on"`
+- `automaticHttpsRewrites: "on"`
+- `ssl: "strict"`
+- `universalSsl: "on"`.
+
+Without these settings, our traffic will be available over HTTP and ... come on, it's 2022 already.
+
+```typescript
+const linkZoneNames = ["pulumi.tv"];
```

Comment:
We should also point out that this goes into `index.js`. (Just 'cause it's not clear otherwise.)

Also, is the assumption that this domain name already exists and is being managed by Cloudflare? I have a domain registered with Cloudflare, for example, but when I use that domain here, I get an error about the zone already existing. However, when I use `getZone()` or `getZoneOutput()`, I get validation errors, and so far haven't figured out how to get past them:

```
warning: provider config warning: Argument is deprecated
    error: Error: Invoke of 'cloudflare:index/getZone:getZone' failed: Invalid combination of arguments: "zone_id": one of `name,zone_id` must be specified (); Invalid combination of arguments: "name": one of `name,zone_id` must be specified ()
```
Bear in mind I haven't actually user CloudFlare much, but nothing I've tried has worked, so feels like others may run into similar questions/problems.